Flake: TestPostgresDatastore/postgres-16-head-/TestRevisionGC

```
=== RUN   TestPostgresDatastore/postgres-16-head-/TestRevisionGC
    revisions.go:166: 
        	Error Trace:	/Users/vroldanbet/git/spicedb/pkg/datastore/test/revisions.go:166
        	            				/Users/vroldanbet/git/spicedb/pkg/datastore/test/datastore.go:104
        	Error:      	An error is expected but got nil.
        	Test:       	TestPostgresDatastore/postgres-16-head-/TestRevisionGC
        	Messages:   	expected revision head-1 to be outside GC Window
        --- FAIL: TestPostgresDatastore/postgres-16-head-/TestRevisionGC (1.13s)
```

--- a/internal/datastore/common/gc.go
+++ b/internal/datastore/common/gc.go
@@ -73,6 +73,10 @@ func RegisterGCMetrics() error {
 // GarbageCollector represents any datastore that supports external garbage
 // collection.
 type GarbageCollector interface {
+	HasGCRun() bool
+	MarkGCCompleted()
+	ResetGCCompleted()
+
 	ReadyState(context.Context) (datastore.ReadyState, error)
 	Now(context.Context) (time.Time, error)
 	TxIDBefore(context.Context, time.Time) (datastore.Revision, error)
@@ -192,5 +196,6 @@ func RunGarbageCollection(gc GarbageCollector, window, timeout time.Duration) er
 	gcRelationshipsCounter.Add(float64(collected.Relationships))
 	gcTransactionsCounter.Add(float64(collected.Transactions))
 	gcNamespacesCounter.Add(float64(collected.Namespaces))
+	gc.MarkGCCompleted()
 	return nil
 }

--- a/internal/datastore/common/gc_test.go
+++ b/internal/datastore/common/gc_test.go
@@ -31,6 +31,16 @@ func (t testGC) DeleteBeforeTx(_ context.Context, _ datastore.Revision) (Deletio
 	return DeletionCounts{}, fmt.Errorf("hi")
 }
 
+func (t testGC) HasGCRun() bool {
+	return true
+}
+
+func (t testGC) MarkGCCompleted() {
+}
+
+func (t testGC) ResetGCCompleted() {
+}
+
 func TestGCFailureBackoff(t *testing.T) {
 	localCounter := prometheus.NewCounter(gcFailureCounterConfig)
 	reg := prometheus.NewRegistry()

--- a/internal/datastore/mysql/datastore.go
+++ b/internal/datastore/mysql/datastore.go
@@ -7,6 +7,7 @@ import (
 	"fmt"
 	"math"
 	"strconv"
+	"sync/atomic"
 	"time"
 
 	sq "github.com/Masterminds/squirrel"
@@ -431,6 +432,7 @@ type Datastore struct {
 	gcGroup  *errgroup.Group
 	gcCtx    context.Context
 	cancelGc context.CancelFunc
+	gcHasRun atomic.Bool
 
 	createTxn     string
 	createBaseTxn string

--- a/internal/datastore/mysql/gc.go
+++ b/internal/datastore/mysql/gc.go
@@ -16,6 +16,18 @@ import (
 
 var _ common.GarbageCollector = (*Datastore)(nil)
 
+func (mds *Datastore) HasGCRun() bool {
+	return mds.gcHasRun.Load()
+}
+
+func (mds *Datastore) MarkGCCompleted() {
+	mds.gcHasRun.Store(true)
+}
+
+func (mds *Datastore) ResetGCCompleted() {
+	mds.gcHasRun.Store(false)
+}
+
 // TODO (@vroldanbet) dupe from postgres datastore - need to refactor
 func (mds *Datastore) Now(ctx context.Context) (time.Time, error) {
 	// Retrieve the `now` time from the database.

--- a/internal/datastore/postgres/gc.go
+++ b/internal/datastore/postgres/gc.go
@@ -31,6 +31,18 @@ var (
 	transactionPKCols = []string{colXID}
 )
 
+func (pgd *pgDatastore) HasGCRun() bool {
+	return pgd.gcHasRun.Load()
+}
+
+func (pgd *pgDatastore) MarkGCCompleted() {
+	pgd.gcHasRun.Store(true)
+}
+
+func (pgd *pgDatastore) ResetGCCompleted() {
+	pgd.gcHasRun.Store(false)
+}
+
 func (pgd *pgDatastore) Now(ctx context.Context) (time.Time, error) {
 	// Retrieve the `now` time from the database.
 	nowSQL, nowArgs, err := getNow.ToSql()

--- a/internal/datastore/postgres/postgres.go
+++ b/internal/datastore/postgres/postgres.go
@@ -5,6 +5,7 @@ import (
 	dbsql "database/sql"
 	"errors"
 	"fmt"
+	"sync/atomic"
 	"time"
 
 	"github.com/IBM/pgxpoolprometheus"
@@ -295,6 +296,7 @@ type pgDatastore struct {
 	gcGroup  *errgroup.Group
 	gcCtx    context.Context
 	cancelGc context.CancelFunc
+	gcHasRun atomic.Bool
 }
 
 func (pgd *pgDatastore) SnapshotReader(revRaw datastore.Revision) datastore.Reader {

--- a/pkg/datastore/test/revisions.go
+++ b/pkg/datastore/test/revisions.go
@@ -124,8 +124,14 @@ func RevisionGCTest(t *testing.T, tester DatastoreTester) {
 	require.NoError(err)
 	require.NoError(ds.CheckRevision(ctx, head), "expected head revision to be valid in GC Window")
 
-	// wait to make sure GC kicks in
-	time.Sleep(400 * time.Millisecond)
+	// Make sure GC kicks in after the window.
+	time.Sleep(300 * time.Millisecond)
+
+	gcable, ok := ds.(common.GarbageCollector)
+	if ok {
+		gcable.ResetGCCompleted()
+		require.Eventually(func() bool { return gcable.HasGCRun() }, 5*time.Second, 50*time.Millisecond, "GC was never run as expected")
+	}
 
 	// FIXME currently the various datastores behave differently when a revision was requested and GC Window elapses.
 	// this is due to the fact MySQL and PostgreSQL implement revisions as a snapshot, while CRDB, Spanner and MemDB

